# Exploratory Analysis

This notebook contains the statistical analysis, distributions, groupby calculations, percentages, relationships, and findings for Q1–Q10.


### Q1. What is the distribution of collision severity in Great Britain?

In [46]:
# Calculate the number of collisions in each severity category and order them by frequency.
severity_counts = (
    cleaned_df["collision_severity_label"]
    .value_counts()
)

severity_percent = (
    severity_counts / severity_counts.sum()
) * 100

severity_summary = pd.DataFrame({
    "Collision Severity": severity_counts.index,
    "Number of Collisions": severity_counts.values,
    "Percentage": severity_percent.values.round(2)
})

print(severity_summary)

  Collision Severity  Number of Collisions  Percentage
0             Slight                 74881       73.76
1            Serious                 25191       24.81
2              Fatal                  1453        1.43


# Q2. Which speed limits are associated with more severe collisions?

In [48]:
# Create a severity frequency table for each speed-limit category.
speed_severity = pd.crosstab(
    cleaned_df["speed_limit"],
    cleaned_df["collision_severity_label"]
)

print(speed_severity)

collision_severity_label  Fatal  Serious  Slight
speed_limit                                     
20                          103     4378   15564
30                          428    11895   37350
40                          164     2259    6463
50                          114     1250    3248
60                          470     4153    8086
70                          174     1256    4170


In [49]:
# Convert the speed-limit severity counts into percentages within each speed limit.
speed_severity_percent = pd.crosstab(
    cleaned_df["speed_limit"],
    cleaned_df["collision_severity_label"],
    normalize="index"
) * 100

print(speed_severity_percent.round(2))

collision_severity_label  Fatal  Serious  Slight
speed_limit                                     
20                         0.51    21.84   77.65
30                         0.86    23.95   75.19
40                         1.85    25.42   72.73
50                         2.47    27.10   70.42
60                         3.70    32.68   63.62
70                         3.11    22.43   74.46


In [51]:
# Calculate the combined Serious + Fatal percentage for each speed limit.
severe_rate = speed_severity_percent.copy()

severe_rate["Serious_or_Fatal"] = (
    severe_rate["Serious"] +
    severe_rate["Fatal"]
)

print(
    severe_rate[["Serious_or_Fatal"]].round(2)
)

collision_severity_label  Serious_or_Fatal
speed_limit                               
20                                   22.35
30                                   24.81
40                                   27.27
50                                   29.58
60                                   36.38
70                                   25.54


## Q3. How do weather and road surface conditions relate to collision severity?


In [53]:
# Create a severity frequency table for each weather condition.
weather_severity = pd.crosstab(
    cleaned_df["weather_conditions_label"],
    cleaned_df["collision_severity_label"]
)

print(weather_severity)

collision_severity_label  Fatal  Serious  Slight
weather_conditions_label                        
Fine + high winds            17      213     485
Fine no high winds         1248    21088   61069
Fog or mist                  11       99     232
Other                        21      668    2625
Raining + high winds         13      234     633
Raining no high winds       138     2528    7564
Snowing + high winds          0        2      17
Snowing no high winds         2       65     168
Unknown                       3      294    2088


In [54]:
# Convert weather-condition severity counts into percentages within each weather category.
weather_severity_percent = pd.crosstab(
    cleaned_df["weather_conditions_label"],
    cleaned_df["collision_severity_label"],
    normalize="index"
) * 100

weather_severity_percent = weather_severity_percent[
    ["Slight", "Serious", "Fatal"]
]

print(weather_severity_percent.round(2))

collision_severity_label  Slight  Serious  Fatal
weather_conditions_label                        
Fine + high winds          67.83    29.79   2.38
Fine no high winds         73.22    25.28   1.50
Fog or mist                67.84    28.95   3.22
Other                      79.21    20.16   0.63
Raining + high winds       71.93    26.59   1.48
Raining no high winds      73.94    24.71   1.35
Snowing + high winds       89.47    10.53   0.00
Snowing no high winds      71.49    27.66   0.85
Unknown                    87.55    12.33   0.13


In [55]:
# Create a severity frequency table for each road-surface condition.
surface_severity = pd.crosstab(
    cleaned_df["road_surface_conditions_label"],
    cleaned_df["collision_severity_label"]
)

print(surface_severity)

collision_severity_label       Fatal  Serious  Slight
road_surface_conditions_label                        
Data missing or out of range       2       65     567
Dry                             1078    18851   55750
Flood over 3cm deep                4       45     102
Frost or ice                      18      294     949
Snow                               3       40     103
Unknown                            0      121    1221
Wet or damp                      348     5775   16189


In [56]:
# Convert road-surface severity counts into percentages within each surface category.
surface_severity_percent = pd.crosstab(
    cleaned_df["road_surface_conditions_label"],
    cleaned_df["collision_severity_label"],
    normalize="index"
) * 100

surface_severity_percent = surface_severity_percent[
    ["Slight", "Serious", "Fatal"]
]

print(surface_severity_percent.round(2))

collision_severity_label       Slight  Serious  Fatal
road_surface_conditions_label                        
Data missing or out of range    89.43    10.25   0.32
Dry                             73.67    24.91   1.42
Flood over 3cm deep             67.55    29.80   2.65
Frost or ice                    75.26    23.31   1.43
Snow                            70.55    27.40   2.05
Unknown                         90.98     9.02   0.00
Wet or damp                     72.56    25.88   1.56


In [59]:
# Calculate the Serious + Fatal rate for each weather condition.
weather_severe_rate = (
    weather_severity_percent["Serious"] +
    weather_severity_percent["Fatal"]
)

print(
    weather_severe_rate
    .sort_values(ascending=False)
    .round(2)
)

weather_conditions_label
Fine + high winds        32.17
Fog or mist              32.16
Snowing no high winds    28.51
Raining + high winds     28.07
Fine no high winds       26.78
Raining no high winds    26.06
Other                    20.79
Unknown                  12.45
Snowing + high winds     10.53
dtype: float64


In [60]:
# Calculate the Serious + Fatal rate for each road-surface condition.
surface_severe_rate = (
    surface_severity_percent["Serious"] +
    surface_severity_percent["Fatal"]
)

print(
    surface_severe_rate
    .sort_values(ascending=False)
    .round(2)
)

road_surface_conditions_label
Flood over 3cm deep             32.45
Snow                            29.45
Wet or damp                     27.44
Dry                             26.33
Frost or ice                    24.74
Data missing or out of range    10.57
Unknown                          9.02
dtype: float64


## Q4. Does collision severity vary between urban and rural areas?


In [63]:
# Count collisions by urban/rural classification and severity.
urban_rural_counts = pd.crosstab(
    cleaned_df["urban_or_rural_area_label"],
    cleaned_df["collision_severity_label"]
)

print(urban_rural_counts)

collision_severity_label   Fatal  Serious  Slight
urban_or_rural_area_label                        
Rural                        932     9682   23759
Unallocated                    0        2       2
Urban                        521    15507   51120


In [64]:
urban_rural_df = cleaned_df[
    cleaned_df["urban_or_rural_area_label"].isin(["Urban", "Rural"])
].copy()

print(urban_rural_df["urban_or_rural_area_label"].value_counts())

urban_or_rural_area_label
Urban    67148
Rural    34373
Name: count, dtype: int64


In [65]:
# Calculate severity percentages within each urban/rural category.
urban_rural_percent = pd.crosstab(
    urban_rural_df["urban_or_rural_area_label"],
    urban_rural_df["collision_severity_label"],
    normalize="index"
) * 100

urban_rural_percent = urban_rural_percent[
    ["Slight", "Serious", "Fatal"]
]

print(urban_rural_percent.round(2))

collision_severity_label   Slight  Serious  Fatal
urban_or_rural_area_label                        
Rural                       69.12    28.17   2.71
Urban                       76.13    23.09   0.78


In [66]:
# Calculate the combined Serious + Fatal rate for urban and rural areas.
urban_rural_severe_rate = (
    urban_rural_percent["Serious"] +
    urban_rural_percent["Fatal"]
)

print(urban_rural_severe_rate.round(2))

urban_or_rural_area_label
Rural    30.88
Urban    23.87
dtype: float64


## Q5. Which road types and junction conditions experience more collisions?

In [69]:
# Calculate road-type collision counts and severity distributions.
road_type_counts = cleaned_df["road_type_label"].value_counts()

road_type_severity = pd.crosstab(
    cleaned_df["road_type_label"],
    cleaned_df["collision_severity_label"]
)

road_type_severity_percent = pd.crosstab(
    cleaned_df["road_type_label"],
    cleaned_df["collision_severity_label"],
    normalize="index"
) * 100

road_type_severity_percent = road_type_severity_percent[["Slight", "Serious", "Fatal"]]

print("Road type collision counts:")
print(road_type_counts)
print("\nRoad type severity percentages:")
print(road_type_severity_percent.round(2))

Road type collision counts:
road_type_label
Single carriageway    74150
Dual carriageway      14610
Roundabout             6795
One way street         2398
Unknown                1953
Slip road              1619
Name: count, dtype: int64

Road type severity percentages:
collision_severity_label  Slight  Serious  Fatal
road_type_label                                 
Dual carriageway           75.44    22.72   1.83
One way street             79.44    20.31   0.25
Roundabout                 80.53    18.84   0.63
Single carriageway         72.07    26.42   1.51
Slip road                  79.99    18.96   1.05
Unknown                    89.30    10.70   0.00


In [70]:
# Count the total number of collisions for each road type.
road_type_counts

road_type_label
Single carriageway    74150
Dual carriageway      14610
Roundabout             6795
One way street         2398
Unknown                1953
Slip road              1619
Name: count, dtype: int64

In [71]:
# Create a severity frequency table for each road type.
road_type_severity

collision_severity_label,Fatal,Serious,Slight
road_type_label,,,
Dual carriageway,268,3320,11022
One way street,6,487,1905
Roundabout,43,1280,5472
Single carriageway,1119,19588,53443
Slip road,17,307,1295
Unknown,0,209,1744


In [72]:
# Convert road-type severity counts into percentages within each road type.
road_type_severity_percent

collision_severity_label,Slight,Serious,Fatal
road_type_label,,,
Dual carriageway,75.441478,22.724162,1.834360
One way street,79.441201,20.308590,0.250209
Roundabout,80.529801,18.837380,0.632818
Single carriageway,72.074174,26.416723,1.509103
Slip road,79.987647,18.962322,1.050031
Unknown,89.298515,10.701485,0.000000


In [75]:
# Count collisions by junction condition.
junction_counts = (
    cleaned_df["junction_detail_label"]
    .value_counts()
)

print(junction_counts)

junction_detail_label
Not at or within 20 metres of junction                45333
T or staggered junction                               28198
Crossroads                                             9456
Unknown (self-reported)                                6762
Data missing or out of range                           6701
Using private drive or entrance                        2144
Junction with more than four arms (not roundabout)     1504
Other junction                                         1427
Name: count, dtype: int64


In [76]:
# Create a severity frequency table for each junction condition.
junction_severity = pd.crosstab(
    cleaned_df["junction_detail_label"],
    cleaned_df["collision_severity_label"]
)

print(junction_severity)

collision_severity_label                            Fatal  Serious  Slight
junction_detail_label                                                     
Crossroads                                             72     2234    7150
Data missing or out of range                           44     1311    5346
Junction with more than four arms (not roundabout)      9      390    1105
Not at or within 20 metres of junction                948    11969   32416
Other junction                                          0       77    1350
T or staggered junction                               277     6983   20938
Unknown (self-reported)                                74     1669    5019
Using private drive or entrance                        29      558    1557


In [77]:
# Convert junction severity counts into percentages within each junction category.
junction_severity_percent = pd.crosstab(
    cleaned_df["junction_detail_label"],
    cleaned_df["collision_severity_label"],
    normalize="index"
) * 100

junction_severity_percent = junction_severity_percent[
    ["Slight", "Serious", "Fatal"]
]

print(junction_severity_percent.round(2))

collision_severity_label                            Slight  Serious  Fatal
junction_detail_label                                                     
Crossroads                                           75.61    23.63   0.76
Data missing or out of range                         79.78    19.56   0.66
Junction with more than four arms (not roundabout)   73.47    25.93   0.60
Not at or within 20 metres of junction               71.51    26.40   2.09
Other junction                                       94.60     5.40   0.00
T or staggered junction                              74.25    24.76   0.98
Unknown (self-reported)                              74.22    24.68   1.09
Using private drive or entrance                      72.62    26.03   1.35


## Q6. How does the number of vehicles involved relate to severity?


In [81]:
# Count collisions by number of vehicles involved.
vehicle_counts = (
    cleaned_df["number_of_vehicles"]
    .value_counts()
    .sort_index()
)

print(vehicle_counts)

number_of_vehicles
1     30841
2     61869
3      6876
4      1331
5       405
6       109
7        58
8        18
9         7
10        5
11        2
12        3
17        1
Name: count, dtype: int64


In [82]:
# Create a severity frequency table for each vehicle count.
vehicle_severity = pd.crosstab(
    cleaned_df["number_of_vehicles"],
    cleaned_df["collision_severity_label"]
)

print(vehicle_severity)

collision_severity_label  Fatal  Serious  Slight
number_of_vehicles                              
1                           648     9474   20719
2                           605    13466   47798
3                           140     1736    5000
4                            34      349     948
5                            18      105     282
6                             4       34      71
7                             0       19      39
8                             4        2      12
9                             0        1       6
10                            0        3       2
11                            0        1       1
12                            0        1       2
17                            0        0       1


In [83]:
# Convert vehicle-count severity counts into percentages within each vehicle-count category.
vehicle_severity_percent = pd.crosstab(
    cleaned_df["number_of_vehicles"],
    cleaned_df["collision_severity_label"],
    normalize="index"
) * 100

vehicle_severity_percent = vehicle_severity_percent[
    ["Slight", "Serious", "Fatal"]
]

print(vehicle_severity_percent.round(2))

collision_severity_label  Slight  Serious  Fatal
number_of_vehicles                              
1                          67.18    30.72   2.10
2                          77.26    21.77   0.98
3                          72.72    25.25   2.04
4                          71.22    26.22   2.55
5                          69.63    25.93   4.44
6                          65.14    31.19   3.67
7                          67.24    32.76   0.00
8                          66.67    11.11  22.22
9                          85.71    14.29   0.00
10                         40.00    60.00   0.00
11                         50.00    50.00   0.00
12                         66.67    33.33   0.00
17                        100.00     0.00   0.00


In [84]:
# Calculate the combined Serious + Fatal rate for each vehicle-count category.
vehicle_severe_rate = (
    vehicle_severity_percent["Serious"] +
    vehicle_severity_percent["Fatal"]
)

print(
    vehicle_severe_rate.round(2)
)

number_of_vehicles
1     32.82
2     22.74
3     27.28
4     28.78
5     30.37
6     34.86
7     32.76
8     33.33
9     14.29
10    60.00
11    50.00
12    33.33
17     0.00
dtype: float64


In [85]:
# Group high vehicle counts together so sparse categories can be compared more reliably.
cleaned_df["vehicle_group"] = np.where(
    cleaned_df["number_of_vehicles"] >= 5,
    "5+ vehicles",
    cleaned_df["number_of_vehicles"].astype(str)
)

In [86]:
print(
    cleaned_df["vehicle_group"].value_counts()
)

vehicle_group
2              61869
1              30841
3               6876
4               1331
5+ vehicles      608
Name: count, dtype: int64


In [87]:
# Create a severity frequency table for the grouped vehicle categories.
vehicle_group_severity = pd.crosstab(
    cleaned_df["vehicle_group"],
    cleaned_df["collision_severity_label"],
    normalize="index"
) * 100

vehicle_group_severity = vehicle_group_severity[
    ["Slight", "Serious", "Fatal"]
]

print(vehicle_group_severity.round(2))

collision_severity_label  Slight  Serious  Fatal
vehicle_group                                   
1                          67.18    30.72   2.10
2                          77.26    21.77   0.98
3                          72.72    25.25   2.04
4                          71.22    26.22   2.55
5+ vehicles                68.42    27.30   4.28


In [88]:
# Calculate the combined Serious + Fatal rate for each grouped vehicle category.
vehicle_group_severe_rate = (
    vehicle_group_severity["Serious"] +
    vehicle_group_severity["Fatal"]
)

print(
    vehicle_group_severe_rate.round(2)
)

vehicle_group
1              32.82
2              22.74
3              27.28
4              28.78
5+ vehicles    31.58
dtype: float64


In [89]:
print(
    cleaned_df["vehicle_group"].value_counts()
)

vehicle_group
2              61869
1              30841
3               6876
4               1331
5+ vehicles      608
Name: count, dtype: int64


In [90]:
print(
    vehicle_group_severity.round(2)
)

collision_severity_label  Slight  Serious  Fatal
vehicle_group                                   
1                          67.18    30.72   2.10
2                          77.26    21.77   0.98
3                          72.72    25.25   2.04
4                          71.22    26.22   2.55
5+ vehicles                68.42    27.30   4.28


## Q7. At what times and days do collisions occur most frequently?


In [93]:
# Define the logical weekday order for consistent plotting.
day_order = [
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
    "Sunday"
]

day_counts = (
    cleaned_df["day_of_week_label"]
    .value_counts()
    .reindex(day_order)
)

print(day_counts)

day_of_week_label
Monday       14309
Tuesday      14880
Wednesday    15369
Thursday     15304
Friday       16811
Saturday     13655
Sunday       11197
Name: count, dtype: int64


In [94]:
# Define the logical time-period order for consistent plotting.
time_period_order = [
    "Morning",
    "Afternoon",
    "Evening",
    "Night"
]

time_period_counts = (
    cleaned_df["time_period"]
    .value_counts()
    .reindex(time_period_order)
)

print(time_period_counts)

time_period
Morning      26843
Afternoon    44365
Evening      24734
Night         5583
Name: count, dtype: int64


In [95]:
# Count collisions for each hour of the day.
hour_counts = (
    cleaned_df["hour"]
    .value_counts()
    .sort_index()
)

print(hour_counts)

hour
0     1525
1     1032
2      789
3      651
4      634
5      952
6     1854
7     4248
8     6716
9     4493
10    4463
11    5069
12    5815
13    6148
14    6466
15    8381
16    8678
17    8877
18    6991
19    5356
20    4022
21    3368
22    2815
23    2182
Name: count, dtype: int64


In [96]:
peak_hour = hour_counts.idxmax()
peak_hour_count = hour_counts.max()

print("Peak hour:", peak_hour)
print("Number of collisions:", peak_hour_count)

Peak hour: 17
Number of collisions: 8877


In [100]:
# Identify the day and time period with the highest collision count.
peak_day_time = day_time_counts.stack().idxmax()
peak_count = day_time_counts.stack().max()

print("Peak day and time period:", peak_day_time)
print("Number of collisions:", peak_count)

Peak day and time period: ('Friday', 'Afternoon')
Number of collisions: 7487


## Q8. Is there a relationship between number of casualties and collision severity?


In [101]:
# Calculate descriptive statistics for casualties within each severity category.
casualty_stats = cleaned_df.groupby(
    "collision_severity_label"
)["number_of_casualties"].agg(
    ["count", "mean", "median", "min", "max", "std"]
)

print(casualty_stats)

                          count      mean  median  min  max       std
collision_severity_label                                             
Fatal                      1453  1.636614     1.0    1   25  1.414246
Serious                   25191  1.342225     1.0    1  142  1.230109
Slight                    74881  1.224516     1.0    1   12  0.591760


In [105]:
# Calculate the average number of casualties for each collision severity.
mean_casualties = (
    cleaned_df.groupby("collision_severity_label")[
        "number_of_casualties"
    ].mean()
    .reindex(["Slight", "Serious", "Fatal"])
)

print(mean_casualties)

collision_severity_label
Slight     1.224516
Serious    1.342225
Fatal      1.636614
Name: number_of_casualties, dtype: float64


## Q9. Which locations/areas have higher collision concentrations?


In [107]:
# Count collisions for each local authority district.
area_counts = (
    cleaned_df["local_authority_ons_district"]
    .value_counts()
)

print(area_counts.head(15))

local_authority_ons_district
E08000025    2307
E08000035    1395
E06000065    1159
E09000033    1133
E06000054    1027
E08000032     956
E09000030     949
E09000022     938
E09000008     937
E09000028     924
E06000052     920
E09000010     861
E06000066     856
E08000012     853
E08000039     838
Name: count, dtype: int64


In [108]:
# Map selected local-authority district codes to readable names.
area_name_mapping = {
    "E08000025": "Birmingham",
    "E08000035": "Leeds",
    "E06000065": "North Yorkshire",
    "E09000033": "Westminster",
    "E06000054": "Wiltshire",
    "E08000032": "Bradford",
    "E09000030": "Tower Hamlets",
    "E09000022": "Lambeth",
    "E09000008": "Croydon",
    "E09000028": "Southwark",
    "E06000052": "Cornwall",
    "E09000010": "Enfield",
    "E06000066": "Somerset",
    "E08000012": "Liverpool",
    "E08000039": "Sheffield"
}

In [109]:
# Add readable local-authority names for easier interpretation of the analysis.
cleaned_df["local_authority_name"] = (
    cleaned_df["local_authority_ons_district"]
    .map(area_name_mapping)
    .fillna(cleaned_df["local_authority_ons_district"])
)

In [111]:
print(area_counts.head(10))

local_authority_name
Birmingham         2307
Leeds              1395
North Yorkshire    1159
Westminster        1133
Wiltshire          1027
Bradford            956
Tower Hamlets       949
Lambeth             938
Croydon             937
Southwark           924
Name: count, dtype: int64


In [112]:
print(cleaned_df[["latitude", "longitude"]].isnull().sum())

latitude     0
longitude    0
dtype: int64


In [114]:
# Filter the dataset to Serious and Fatal collisions for geographic comparison.
serious_fatal_df = cleaned_df[
    cleaned_df["collision_severity_label"].isin(["Serious", "Fatal"])
].copy()

## Q10. Which combinations of road, environmental, traffic, and temporal conditions are frequently observed in serious and fatal collisions?


In [116]:
# Filter the dataset to Serious and Fatal collisions only

serious_fatal_df = cleaned_df[
    cleaned_df["collision_severity_label"].isin(["Serious", "Fatal"])
].copy()

print("Total Serious/Fatal collisions:", len(serious_fatal_df))

print("\nSeverity distribution:")
print(
    serious_fatal_df["collision_severity_label"]
    .value_counts()
)

Total Serious/Fatal collisions: 26644

Severity distribution:
collision_severity_label
Serious    25191
Fatal       1453
Name: count, dtype: int64


In [117]:
# Create combinations of road, environmental, traffic and temporal conditions

q10_combinations = (
    serious_fatal_df
    .groupby([
        "road_type_label",
        "junction_detail_label",
        "weather_conditions_label",
        "road_surface_conditions_label",
        "speed_limit",
        "time_period"
    ])
    .size()
    .reset_index(name="serious_fatal_collisions")
)

# Sort by frequency
q10_combinations = q10_combinations.sort_values(
    "serious_fatal_collisions",
    ascending=False
)

q10_combinations.head(20)

,road_type_label,junction_detail_label,weather_conditions_label,road_surface_conditions_label,speed_limit,time_period,serious_fatal_collisions
1116,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,30,Afternoon,1460
1427,Single carriageway,T or staggered junction,Fine no high winds,Dry,30,Afternoon,1256
1128,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,60,Afternoon,884
1117,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,30,Evening,842
1428,Single carriageway,T or staggered junction,Fine no high winds,Dry,30,Evening,676
1118,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,30,Morning,666
1429,Single carriageway,T or staggered junction,Fine no high winds,Dry,30,Morning,663
1129,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,60,Evening,498
1423,Single carriageway,T or staggered junction,Fine no high winds,Dry,20,Afternoon,481
1112,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,20,Afternoon,454


In [118]:
# Keep only combinations that occur at least 20 times
# among Serious/Fatal collisions

q10_frequent = q10_combinations[
    q10_combinations["serious_fatal_collisions"] >= 20
].copy()

print(
    "Number of frequent combinations:",
    len(q10_frequent)
)

q10_frequent.head(20)

Number of frequent combinations: 196


,road_type_label,junction_detail_label,weather_conditions_label,road_surface_conditions_label,speed_limit,time_period,serious_fatal_collisions
1116,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,30,Afternoon,1460
1427,Single carriageway,T or staggered junction,Fine no high winds,Dry,30,Afternoon,1256
1128,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,60,Afternoon,884
1117,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,30,Evening,842
1428,Single carriageway,T or staggered junction,Fine no high winds,Dry,30,Evening,676
1118,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,30,Morning,666
1429,Single carriageway,T or staggered junction,Fine no high winds,Dry,30,Morning,663
1129,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,60,Evening,498
1423,Single carriageway,T or staggered junction,Fine no high winds,Dry,20,Afternoon,481
1112,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,20,Afternoon,454


In [119]:
# Top 15 most frequently observed combinations

top_q10 = q10_frequent.head(15).copy()

print(
    top_q10[
        [
            "road_type_label",
            "junction_detail_label",
            "weather_conditions_label",
            "road_surface_conditions_label",
            "speed_limit",
            "time_period",
            "serious_fatal_collisions"
        ]
    ].to_string(index=False)
)

   road_type_label                  junction_detail_label weather_conditions_label road_surface_conditions_label  speed_limit time_period  serious_fatal_collisions
Single carriageway Not at or within 20 metres of junction       Fine no high winds                           Dry           30   Afternoon                      1460
Single carriageway                T or staggered junction       Fine no high winds                           Dry           30   Afternoon                      1256
Single carriageway Not at or within 20 metres of junction       Fine no high winds                           Dry           60   Afternoon                       884
Single carriageway Not at or within 20 metres of junction       Fine no high winds                           Dry           30     Evening                       842
Single carriageway                T or staggered junction       Fine no high winds                           Dry           30     Evening                       676
Single carriagew

In [120]:
# Calculate the percentage of all Serious/Fatal collisions
# represented by each condition combination

total_serious_fatal = len(serious_fatal_df)

top_q10["percentage_of_serious_fatal"] = (
    top_q10["serious_fatal_collisions"]
    / total_serious_fatal
    * 100
)

top_q10[
    [
        "road_type_label",
        "junction_detail_label",
        "weather_conditions_label",
        "road_surface_conditions_label",
        "speed_limit",
        "time_period",
        "serious_fatal_collisions",
        "percentage_of_serious_fatal"
    ]
]

,road_type_label,junction_detail_label,weather_conditions_label,road_surface_conditions_label,speed_limit,time_period,serious_fatal_collisions,percentage_of_serious_fatal
1116,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,30,Afternoon,1460,5.479658
1427,Single carriageway,T or staggered junction,Fine no high winds,Dry,30,Afternoon,1256,4.714007
1128,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,60,Afternoon,884,3.317820
1117,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,30,Evening,842,3.160186
1428,Single carriageway,T or staggered junction,Fine no high winds,Dry,30,Evening,676,2.537157
1118,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,30,Morning,666,2.499625
1429,Single carriageway,T or staggered junction,Fine no high winds,Dry,30,Morning,663,2.488365
1129,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,60,Evening,498,1.869089
1423,Single carriageway,T or staggered junction,Fine no high winds,Dry,20,Afternoon,481,1.805284
1112,Single carriageway,Not at or within 20 metres of junction,Fine no high winds,Dry,20,Afternoon,454,1.703948


In [121]:
# Create a readable label for each condition combination

top_q10["condition_combination"] = (
    top_q10["road_type_label"].astype(str)
    + " | "
    + top_q10["junction_detail_label"].astype(str)
    + " | "
    + top_q10["weather_conditions_label"].astype(str)
    + " | "
    + top_q10["road_surface_conditions_label"].astype(str)
    + " | "
    + top_q10["speed_limit"].astype(str)
    + " mph | "
    + top_q10["time_period"].astype(str)
)

top_q10[
    [
        "condition_combination",
        "serious_fatal_collisions",
        "percentage_of_serious_fatal"
    ]
]

,condition_combination,serious_fatal_collisions,percentage_of_serious_fatal
1116,Single carriageway | Not at or within 20 metre...,1460,5.479658
1427,Single carriageway | T or staggered junction |...,1256,4.714007
1128,Single carriageway | Not at or within 20 metre...,884,3.317820
1117,Single carriageway | Not at or within 20 metre...,842,3.160186
1428,Single carriageway | T or staggered junction |...,676,2.537157
1118,Single carriageway | Not at or within 20 metre...,666,2.499625
1429,Single carriageway | T or staggered junction |...,663,2.488365
1129,Single carriageway | Not at or within 20 metre...,498,1.869089
1423,Single carriageway | T or staggered junction |...,481,1.805284
1112,Single carriageway | Not at or within 20 metre...,454,1.703948


In [123]:
# Analyze number of vehicles involved in Serious/Fatal collisions

vehicle_frequency = (
    serious_fatal_df["number_of_vehicles"]
    .value_counts()
    .sort_index()
)

print(vehicle_frequency)

number_of_vehicles
1     10122
2     14071
3      1876
4       383
5       123
6        38
7        19
8         6
9         1
10        3
11        1
12        1
Name: count, dtype: int64


In [124]:
# Group rare high vehicle counts into 5+ vehicles

serious_fatal_df["vehicle_group"] = np.where(
    serious_fatal_df["number_of_vehicles"] >= 5,
    "5+ vehicles",
    serious_fatal_df["number_of_vehicles"].astype(str)
)

vehicle_group_frequency = (
    serious_fatal_df["vehicle_group"]
    .value_counts()
)

print(vehicle_group_frequency)

vehicle_group
2              14071
1              10122
3               1876
4                383
5+ vehicles      192
Name: count, dtype: int64


In [126]:
# Identify the most frequent day and time-period combinations
# among Serious/Fatal collisions

temporal_frequency = (
    serious_fatal_df
    .groupby([
        "day_of_week_label",
        "time_period"
    ])
    .size()
    .reset_index(
        name="serious_fatal_collisions"
    )
    .sort_values(
        "serious_fatal_collisions",
        ascending=False
    )
)

print(
    temporal_frequency.head(10).to_string(index=False)
)

day_of_week_label time_period  serious_fatal_collisions
           Friday   Afternoon                      1869
         Thursday   Afternoon                      1695
        Wednesday   Afternoon                      1676
          Tuesday   Afternoon                      1615
         Saturday   Afternoon                      1607
           Monday   Afternoon                      1597
           Sunday   Afternoon                      1345
           Friday     Evening                      1254
         Saturday     Evening                      1142
        Wednesday     Morning                      1042


In [128]:
# Perform a final quality check before saving the cleaned dataset.
print("===== FINAL DATASET CHECK =====")

print("Shape:", cleaned_df.shape)

print("\nMissing values:")
print(cleaned_df.isnull().sum()[cleaned_df.isnull().sum() > 0])

print("\nDuplicate rows:", cleaned_df.duplicated().sum())

print("\nTarget distribution:")
print(cleaned_df["collision_severity_label"].value_counts())

print("\nData types:")
print(cleaned_df.dtypes)

===== FINAL DATASET CHECK =====
Shape: (101525, 59)

Missing values:
Series([], dtype: int64)

Duplicate rows: 0

Target distribution:
collision_severity_label
Slight     74881
Serious    25191
Fatal       1453
Name: count, dtype: int64

Data types:
collision_index                                                str
collision_year                                               int64
collision_ref_no                                               str
location_easting_osgr                                      float64
location_northing_osgr                                     float64
longitude                                                  float64
latitude                                                   float64
police_force                                                 int64
collision_severity                                           int64
number_of_vehicles                                           int64
number_of_casualties                                         int64
date         

In [130]:
# Save the final cleaned dataset so it can be reused without modifying the raw data.
cleaned_df.to_csv(
    "Great_Britain_Road_Collisions_2025_Cleaned.csv",
    index=False
)

print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


# Overall EDA Findings

1. Slight collisions dominate the dataset, representing approximately 73.8% of all collisions.
2. 60 mph roads show the highest proportion of serious/fatal collisions.
3. Rural collisions have a higher serious/fatal proportion than urban collisions.
4. Single carriageways account for the largest number of collisions.
5. Friday is the busiest collision day, while 17:00 is the peak collision hour.
6. Fatal collisions have the highest average number of casualties.
7. Birmingham has the highest number of recorded collisions among local authority areas.
8. Higher-risk combinations are concentrated around higher speed limits, rural environments, and certain weather/surface conditions.
9. Weather and road-surface categories with very small sample sizes should be interpreted cautiously.
10. Collision severity and casualty count have an inherent relationship in the dataset and should not be interpreted as an independent causal relationship.